File: python/aufgabe01.py
import argparse
import numpy as np
import time

class FFT_Analyzer:
    def __init__(self, filename, blocksize, shift, threshold):
        self.filename = filename
        self.blocksize = max(min(512, int(blocksize)), 64)
        self.shift = max(min(self.blocksize, int(shift)), 1)
        self.threshold = int(threshold)
    
    def get_amplitude_mean(self):
        data = np.fromfile(self.filename, dtype=np.int16)
        samples = len(data) // 2
        normalized_data_left = (data[::2] / (2 ** 15)).astype(np.float32)

        bins = np.zeros(self.blocksize // 2)
        offset = 0

        while offset < samples:
            if (offset + self.blocksize) > samples:
                break
            
            block = normalized_data_left[offset:offset+self.blocksize]
            fft_data = np.fft.fft(block)
            amplitude = np.abs(fft_data[:self.blocksize//2])
            bins += amplitude
            offset += self.shift

        bins = bins / (samples // self.shift)        
        bins_db = 20 * np.log10(bins)
        return bins_db
        

def main(args):
    fft_analyzer = FFT_Analyzer(args.filename, args.blocksize, args.shift, args.threshold)
    start_time = time.time()
    result = fft_analyzer.get_amplitude_mean()
    end_time = time.time()
    execution_time = end_time - start_time
    for x in range(len(result)):
        if(result[x] > fft_analyzer.threshold):
            print(f"{x*(44100//(fft_analyzer.blocksize/2))}Hz {result[x]}")
        
    print(f"Execution time: {execution_time} seconds")

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('filename')
    parser.add_argument('blocksize')
    parser.add_argument('shift')
    parser.add_argument('threshold')
    
    args = parser.parse_args()
    main(args)

File: c/aufgabe01.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <sys/time.h>
#include "fftw3.h"

#define MAX(a,b) ((a) > (b) ? (a) : (b))
#define MIN(a,b) ((a) < (b) ? (a) : (b))

typedef struct {
  char* filename;
  int blocksize;
  int shift;
  int threshold;
} FFT_Analyzer;

FFT_Analyzer* create_fft_analyzer(const char* filename, int blocksize, int shift, int threshold) {
  FFT_Analyzer* analyzer = malloc(sizeof(FFT_Analyzer));
  analyzer->filename = strdup(filename);
  analyzer->blocksize = MAX(MIN(512, blocksize), 64);
  analyzer->shift = MAX(MIN(analyzer->blocksize, shift), 1);
  analyzer->threshold = threshold;
  return analyzer;
}

void destroy_fft_analyzer(FFT_Analyzer* analyzer) {
  free(analyzer->filename);
  free(analyzer);
}

double* get_amplitude_mean(FFT_Analyzer* analyzer) {
  FILE* file = fopen(analyzer->filename, "rb");
  if (!file) {
    perror("Error opening file");
    return NULL;
  }

  fseek(file, 0, SEEK_END);
  long file_size = ftell(file);
  fseek(file, 0, SEEK_SET);

  int samples = file_size / 4;
  short* data = malloc(file_size);
  fread(data, 2, samples * 2, file);
  fclose(file);

  double* normalized_data_left = malloc(samples * sizeof(double));
  for (int i = 0; i < samples; i++) {
    normalized_data_left[i] = data[i*2] / 32768.0;
  }
  free(data);

  int bins_size = analyzer->blocksize / 2;
  double* bins = calloc(bins_size, sizeof(double));

  fftw_complex* fft_out = fftw_malloc(sizeof(fftw_complex) * (analyzer->blocksize/2 + 1));
  double* fft_in = fftw_malloc(sizeof(double) * analyzer->blocksize);
  fftw_plan plan = fftw_plan_dft_r2c_1d(analyzer->blocksize, fft_in, fft_out, FFTW_PATIENT);

  int offset = 0;
  int count = 0;
  while (offset + analyzer->blocksize <= samples) {
    memcpy(fft_in, normalized_data_left + offset, analyzer->blocksize * sizeof(double));
    fftw_execute(plan);

    for (int i = 0; i < bins_size; i++) {
      double real = fft_out[i][0];
      double imag = fft_out[i][1];
      bins[i] += sqrt(real*real + imag*imag);
    }

    offset += analyzer->shift;
    count++;
  }

  fftw_destroy_plan(plan);
  fftw_free(fft_in);
  fftw_free(fft_out);
  free(normalized_data_left);

  for (int i = 0; i < bins_size; i++) {
    bins[i] /= count;
    bins[i] = 20 * log10(bins[i]);
  }

  return bins;
}

int main(int argc, char* argv[]) {
  if (argc != 5) {
    fprintf(stderr, "Usage: %s <filename> <blocksize> <shift> <threshold>\n", argv[0]);
    return 1;
  }

  FFT_Analyzer* analyzer = create_fft_analyzer(argv[1], atoi(argv[2]), atoi(argv[3]), atoi(argv[4]));

  struct timeval start, end;
  gettimeofday(&start, NULL);
  double* result = get_amplitude_mean(analyzer);
  gettimeofday(&end, NULL);

  if (result) {
    for (int i = 0; i < analyzer->blocksize/2; i++) {
      if(result[i] > analyzer->threshold) {
	printf("%dHz %f\n", i*(44100/(analyzer->blocksize/2)), result[i]);
      }
    }
    printf("\n");
    free(result);
  }



  long seconds = end.tv_sec - start.tv_sec;
  long microseconds = end.tv_usec - start.tv_usec;
  double elapsed_time = seconds + microseconds / 1e6;
  printf("Execution time: %f seconds\n", elapsed_time);

  destroy_fft_analyzer(analyzer);
  return 0;
}
